File: code/go/2020/sept/partition-labels.go
package sept

import "math"

func PartitionLabel(s string) []int {

	var ret []int
	dict := make([]int, 26)
	for i := 0; i < len(s); i++ {
		dict[s[i]-'a'] = i
	}

	start, last := 0, 0
	for i := 0; i < len(s); i++ {
		last = int(math.Max(float64(last), float64(dict[s[i]-'a'])))
		if last == i {
			ret = append(ret, last-start+1)
			start = last + 1
		}
	}
	return ret
}

File: archive/java/src/main/java/cn/az/code/year2020/sept/PartitionLabels.java
package cn.az.code.year2020.sept;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author az
 * @since 09/04/20
 */
public class PartitionLabels {

    public static void main(String[] args) {
        PartitionLabels pl = new PartitionLabels();
        System.out.println(pl.partitionLabels("ababcbacadefegdehijhklij"));
    }

    //traverse the string record the last index of each char.
    //using pointer to record end of the current sub string.

    /**
     *
     * @param S s
     * @return xxx
     */
    public List<Integer> partitionLabels(String S) {
        if (S == null || S.length() == 0) {
            return Collections.emptyList();
        }
        List<Integer> list = new ArrayList<>();
        // record the last index of the each char
        int[] map = new int[26];

        for (int i = 0; i < S.length(); i++) {
            map[S.charAt(i) - 'a'] = i;
        }
        // record the end index of the current sub string
        int last = 0;
        int start = 0;
        for (int i = 0; i < S.length(); i++) {
            last = Math.max(last, map[S.charAt(i) - 'a']);
            if (last == i) {
                list.add(last - start + 1);
                start = last + 1;
            }
        }
        return list;
    }
}
